### PR TITLE
Centralise common lint configurations to workspace manifest

### LINKED ISSUE #5877: Move top-level lints to [workspace.lints] manifest table
### Describe the feature

- In the root /Cargo.toml: add [`[workspace.lints]`](https://doc.rust-lang.org/nightly/cargo/reference/workspaces.html#the-lints-table) with the lints that are in (almost) every crate, and remove said lints from the crates:
  ```rust
  #![warn(missing_debug_implementations, missing_docs, unreachable_pub, rustdoc::all)]
  #![deny(unused_must_use, rust_2018_idioms)]
  ```
- In all package Cargo.toml: add [`[lints] workspace = true`](https://doc.rust-lang.org/nightly/cargo/reference/manifest.html#the-lints-section) after `[package]`
- Run `cargo clippy --workspace` to verify
- It's fine if you have to manually allow some of these in individual crates

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ghost**
This PR resolves #5877 by consolidating the top-level lint configurations into a single `[workspace.lints]` table within the root `Cargo.toml` and reduces redundancy across the various crate-specific `Cargo.toml` files.

In the root `Cargo.toml`, the `[workspace.lints]` table has been introduced, containing the lints that were previously duplicated in nearly every crate. All the specific lint attributes from the individual crate `lib.rs` were subsequently removed, with a `[lints] workspace = true]` line added to every crate `Cargo.toml`.

I ran `cargo clippy --workspace`, and was faced with `unknown_lint` warnings. So I decided to supress them using `unknown_lints = "allow"` in the root `Cargo.toml`. This may be reversed if required.

No other changes were made. Comments were retained in case the author intended for some more documentation to be provided.

### CODE REVIEW

File: Cargo.toml
Comment (tcoratger) [inline Cargo.toml:236]: Is it working? Shouldn't it be something like:

```
[workspace.lints]
rust.missing_debug_implementations = "warn"
rust.missing_docs = "warn"
rust.unreachable_pub = "warn"
rustdoc.all = "warn"
rust.unused_must_use = "deny"
rust.rust_2018_idioms = "deny"
```

Comment (DaniPopes) [reply Cargo.toml:236]: this is fine except for rustdoc which is in lints.rustdoc not lints.rust.rustdoc

File: Cargo.toml
Comment (DaniPopes) [inline Cargo.toml:239]: this shouldn't need quotes

Comment (mx644) [reply Cargo.toml]: Fixed.

File: Cargo.toml
Comment (mx644) [reply Cargo.toml]: Reverted to wotk without the name prefixes. 

> Is it working? Shouldn't it be something like:
> 
> ```
> [workspace.lints]
> rust.missing_debug_implementations = "warn"
> rust.missing_docs = "warn"
> rust.unreachable_pub = "warn"
> rustdoc.all = "warn"
> rust.unused_must_use = "deny"
> rust.rust_2018_idioms = "deny"
> ```

Yes, it's just a name prefix but I removed the changes, as it's a bit light on [documentation](https://rust-lang.github.io/rfcs/3389-manifest-lint.html) around how to use it.

**Review by DaniPopes — APPROVED** [review]
LGTM. Needs rebase before I can run CI and merge this.

**Comment by DaniPopes** [comment]
@mx644 Can you please rebase on main or enable "Allow edits from maintainers"?

**Comment by mx644** [comment]
> @mx644 Can you please rebase on main or enable "Allow edits from maintainers"?

Should be good to go.

**Comment by mx644** [comment]
Sorry @DaniPopes. Need some help here.

**Comment by mx644** [comment]
This fails to compile cleanly due to the fact that we seem to error on multiple functions that have missing documentation,

**Pull request closed without merging**
